app/lib/signed_request.rb

```diff
@@ -0,0 +1,274 @@
+        'signature' => @request.headers['signature'],
+      })
+    end
+
+    def key_id
+      @signature.parameters['keyid']
+    end
+
+    def missing_signature_parameters
+      REQUIRED_PARAMETERS if REQUIRED_PARAMETERS.any? { |p| @signature.parameters[p].blank? }
+    end
+
+    # This method can lie as we only support one specific algorith for now.
+    # But HTTP Message Signatures do not need to specify an algorithm (as
+    # this can be inferred from the key used). Using an unsupported
+    # algorithm will fail anyway further down the line.
+    def algorithm_supported?
+      true
+    end
+
+    def verified?(actor)
+      key = Linzer.new_rsa_v1_5_sha256_public_key(actor.public_key)
+
+      Linzer.verify!(@request.rack_request, key:)
+    rescue Linzer::VerifyError
+      false
+    end
+
+    def verify_signature_strength!
+      raise Mastodon::SignatureVerificationError, 'Mastodon requires the (created) parameter to be signed' if @signature.parameters['created'].blank?
+      raise Mastodon::SignatureVerificationError, 'Mastodon requires the @method and @target-uri derived components to be signed' unless @signature.components.include?('@method') && @signature.components.include?('@target-uri')
+      raise Mastodon::SignatureVerificationError, 'Mastodon requires the Content-Digest header to be signed when doing a POST request' if @request.post? && !signed_headers.include?('content-digest')
+    end
+
+    def verify_body_digest!
+      return unless signed_headers.include?('content-digest')
+      raise Mastodon::SignatureVerificationError, 'Content-Digest header missing' unless @request.headers.key?('content-digest')
+
+      digests = Starry.parse_dictionary(@request.headers['content-digest'])
+      raise Mastodon::SignatureVerificationError, "Mastodon only supports SHA-256 in Content-Digest header. Offered algorithms: #{digests.keys.join(', ')}" unless digests.key?('sha-256')
+
+      received_digest = Base64.strict_encode64(digests['sha-256'].value)
+      return if body_digest == received_digest
+
+      raise Mastodon::SignatureVerificationError, "Invalid Digest value. Computed SHA-256 digest: #{body_digest}; given: #{received_digest}"
+    end
+
+    def created_time
+      Time.at(@signature.parameters['created'].to_i).utc
+    rescue ArgumentError => e
+      raise Mastodon::SignatureVerificationError, "Invalid Date header: #{e.message}"
+    end
+
+    def expires_time
+      Time.at(@signature.parameters['expires'].to_i).utc if @signature.parameters['expires'].present?
+    rescue ArgumentError => e
+      raise Mastodon::SignatureVerificationError, "Invalid Date header: #{e.message}"
+    end
+
+    private
+
+    def request_body
+      @request_body ||= @request.raw_post
+    end
+
+    def signed_headers
+      @signed_headers ||= @signature.components.reject { |c| c.start_with?('@') }
+    end
+
+    def signed_headers_from_request
+      signed_headers.index_with { |h| @request.headers[h] }
+    end
```

Comment:
Is this actually used anywhere?

Reply:
No, it is not. Thanks for pointing this out, I just remove the method.